opensearch/src/test/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateTypeTest.java

```diff
@@ -0,0 +1,220 @@
+import org.opensearch.common.time.DateFormatter;
+import org.opensearch.common.time.FormatNames;
+import org.opensearch.sql.data.type.ExprCoreType;
+import org.opensearch.sql.data.type.ExprType;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+class OpenSearchDateTypeTest {
+  private static final String formatString = "epoch_millis || yyyyMMDD";
+  private static final String defaultFormatString = "";
+
+  private static final String dateFormatString = "date";
+
+  private static final String timeFormatString = "hourMinuteSecond";
+
+  private static final String datetimeFormatString = "basic_date_time";
+
+  private static final OpenSearchDateType defaultDateType =
+      OpenSearchDateType.create(defaultFormatString);
+  private static final OpenSearchDateType dateDateType =
+      OpenSearchDateType.create(dateFormatString);
+  private static final OpenSearchDateType timeDateType =
+      OpenSearchDateType.create(timeFormatString);
+  private static final OpenSearchDateType datetimeDateType =
+      OpenSearchDateType.create(datetimeFormatString);
+
+  @Test
+  public void isCompatible() {
+    // timestamp types is compatible with all date-types
+    assertTrue(TIMESTAMP.isCompatible(defaultDateType));
+    assertTrue(TIMESTAMP.isCompatible(dateDateType));
+    assertTrue(TIMESTAMP.isCompatible(timeDateType));
+    assertTrue(TIMESTAMP.isCompatible(datetimeDateType));
+
+    // datetime
+    assertFalse(DATETIME.isCompatible(defaultDateType));
+    assertTrue(DATETIME.isCompatible(dateDateType));
+    assertTrue(DATETIME.isCompatible(timeDateType));
+    assertFalse(DATETIME.isCompatible(datetimeDateType));
+
+    // time type
+    assertFalse(TIME.isCompatible(defaultDateType));
+    assertFalse(TIME.isCompatible(dateDateType));
+    assertTrue(TIME.isCompatible(timeDateType));
+    assertFalse(TIME.isCompatible(datetimeDateType));
+
+    // date type
+    assertFalse(DATE.isCompatible(defaultDateType));
+    assertTrue(DATE.isCompatible(dateDateType));
+    assertFalse(DATE.isCompatible(timeDateType));
+    assertFalse(DATE.isCompatible(datetimeDateType));
+  }
+
+  // `typeName` and `legacyTypeName` return the same thing for date objects:
+  // https://github.com/opensearch-project/sql/issues/1296
+  @Test
+  public void check_typeName() {
+    // always use the type "DATE"
+    assertEquals("DATE", defaultDateType.typeName());
+    assertEquals("DATE", timeDateType.typeName());
+    assertEquals("DATE", dateDateType.typeName());
+    assertEquals("DATE", datetimeDateType.typeName());
+  }
+
+  @Test
+  public void check_legacyTypeName() {
+    // always use the legacy "DATE" type
+    assertEquals("DATE", defaultDateType.legacyTypeName());
+    assertEquals("DATE", timeDateType.legacyTypeName());
+    assertEquals("DATE", dateDateType.legacyTypeName());
+    assertEquals("DATE", datetimeDateType.legacyTypeName());
+  }
+
+  @Test
+  public void check_exprTypeName() {
+    // always use the legacy "DATE" type
```

Comment:
Should this comment be updated as not all are dates.

Reply:
Updated in https://github.com/Bit-Quill/opensearch-project-sql/pull/268/commits/5f34b391f988e4f2e179a68bce688abc1dc15914